matrix_sdk_appservice/src/warp.rs

```diff
@@ -0,0 +1,198 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{net::ToSocketAddrs, result::Result as StdResult};
+
+use futures::TryFutureExt;
+use matrix_sdk::Bytes;
+use serde::Serialize;
+use warp::{filters::BoxedFilter, path::FullPath, Filter, Rejection, Reply};
+
+use crate::{Appservice, Error, Result};
+
+pub async fn run_server(
+    appservice: Appservice,
+    host: impl Into<String>,
+    port: impl Into<u16>,
+) -> Result<()> {
+    let routes = warp_filter(appservice);
+
+    let mut addr = format!("{}:{}", host.into(), port.into()).to_socket_addrs()?;
+    if let Some(addr) = addr.next() {
+        warp::serve(routes).run(addr).await;
+        Ok(())
+    } else {
+        Err(Error::HostPortToSocketAddrs)
+    }
+}
+
+pub fn warp_filter(appservice: Appservice) -> BoxedFilter<(impl Reply,)> {
+    warp::any()
+        .and(filters::valid_access_token(appservice.registration().hs_token.clone()))
+        .and(filters::transactions(appservice))
+        .or(filters::users())
+        .or(filters::rooms())
+        .recover(handle_rejection)
+        .boxed()
+}
+
+mod filters {
+    use super::*;
+
+    pub fn users() -> BoxedFilter<(impl Reply,)> {
+        warp::get()
+            .and(
+                warp::path!("_matrix" / "app" / "v1" / "users" / String)
+                    // legacy route
+                    .or(warp::path!("users" / String))
+                    .unify(),
+            )
+            .and_then(handlers::user)
+            .boxed()
+    }
+
+    pub fn rooms() -> BoxedFilter<(impl Reply,)> {
+        warp::get()
+            .and(
+                warp::path!("_matrix" / "app" / "v1" / "rooms" / String)
+                    // legacy route
+                    .or(warp::path!("rooms" / String))
+                    .unify(),
+            )
+            .and_then(handlers::room)
+            .boxed()
+    }
+
+    pub fn transactions(appservice: Appservice) -> BoxedFilter<(impl Reply,)> {
+        warp::put()
+            .and(
+                warp::path!("_matrix" / "app" / "v1" / "transactions" / String)
+                    // legacy route
+                    .or(warp::path!("transactions" / String))
+                    .unify(),
+            )
+            .and(with_appservice(appservice))
+            .and(http_request().and_then(|request| async move {
+                let request = crate::transform_legacy_route(request).map_err(Error::from)?;
+                Ok::<http::Request<Bytes>, Rejection>(request)
+            }))
```

Comment:
This should probably wrap all filters given they all have legacy routes and want to be deserialized by ruma